feat(appium): generate declaration files

boneskull opened the pull request:
Configure TS to generate declaration files for `appium` package.

Most of these changes are just types getting added or changed via docstrings; in other places, typechecking has caught problems and I had to refactor code to address them.

- Removed `// @ts-check` pragma from all files in `lib/`, since now `checkJs: true` is in `tsconfig.json`.  It needs to stay in place elsewhere (I think) because only sources in `lib` have declarations generated.
- The `install`, `uninstall`, `update` and `run` functions in the base `ExtensionCommand` class were renamed and prefixed with a `_`; TS does not allow overridden functions with differing signatures.   It no longer exports a default.
- Removed `type` prop of options for `DriverCommand` and `PluginCommand` constructors, respectively (this value is derived from the `ExtensionConfig` `config` parameter instead)
- `AppiumDriver` now extends `DriverCore` instead of `BaseDriver`, since some of its function signatures differ from `BaseDriver`'s.
- Small refactor of `deleteSession` to make typechecker happy.
- `executeCommand` now calls to `BaseDriver` directly if `isUmbrellaCmd` is truthy.  This is one place where inheritance just kind of breaks, since `BaseDriver#executeCommand` returns something entirely different than `AppiumDriver#executeCommand` would otherwise.
- Removed `--delete-dir-on-start` from build command, since it may blast generated declarations.

boneskull committed:
feat(appium): generate declaration files

Configure TS to generate declaration files for `appium` package.

Most of these changes are just types getting added or changed via docstrings; in other places, typechecking has caught problems and I had to refactor code to address them.

- Removed `// @ts-check` pragma from all files in `lib/`, since now `checkJs: true` is in `tsconfig.json`.  It needs to stay in place elsewhere (I think) because only sources in `lib` have declarations generated.
- The `install`, `uninstall`, `update` and `run` functions in the base `ExtensionCommand` class were renamed and prefixed with a `_`; TS does not allow overridden functions with differing signatures.   It no longer exports a default.
- Removed `type` prop of options for `DriverCommand` and `PluginCommand` constructors, respectively (this value is derived from the `ExtensionConfig` `config` parameter instead)
- `AppiumDriver` now extends `DriverCore` instead of `BaseDriver`, since some of its function signatures differ from `BaseDriver`'s.
- Small refactor of `deleteSession` to make typechecker happy.
- `executeCommand` now calls to `BaseDriver` directly if `isUmbrellaCmd` is truthy.  This is one place where inheritance just kind of breaks, since `BaseDriver#executeCommand` returns something entirely different than `AppiumDriver#executeCommand` would otherwise.
- Removed `--delete-dir-on-start` from build command, since it may blast generated declarations.
- Reorganize declarations in `types/`; moved some of them out of `@typedef`s. `types.d.ts` becomes `index.d.ts`
- Note that this contains the declarations for `PluginClass` and `DriverClass`, which correspond to the types of extension "main" classes. This differs from the interfaces provided in `@appium/types` (e.g., `Driver`, `ExternalDriver`) in that these aren't just objects; they are classes with constructors and static members. I may end up moving some of this stuff into `@appium/types`, but I'm not sure how useful it will be.  Will see when I add types to `@appium/fake-driver`.

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16594** <a href="https://app.graphite.dev/github/pr/appium/appium/16594" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #16595** <a href="https://app.graphite.dev/github/pr/appium/appium/16595" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #16596** <a href="https://app.graphite.dev/github/pr/appium/appium/16596" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #16597** <a href="https://app.graphite.dev/github/pr/appium/appium/16597" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

jlipps commented on packages/appium/lib/cli/extension-command.js:
```diff
@@ -307,27 +313,31 @@ export default class ExtensionCommand {
-    return {...appium, pkgName: name, version};
+    // we need to omit `driverName` and `pluginName`. `_.omit()` does the job, but it isn't really
+    // type-safe.  so we're fudging it here.
+    delete appium[/** @type {string} */(`${this.type}Name`)];
```
can ts not detect that `${this.type}Name` is a string?

boneskull replied:
it can now, apparently.

boneskull replied:
pretty sure there was a bug with string template types

boneskull replied:
actually, nevermind. no, it doesn't work. no, I don't know why.

jlipps commented on packages/appium/lib/cli/extension-command.js:
```diff
@@ -307,27 +313,31 @@ export default class ExtensionCommand {
+    // we need to omit `driverName` and `pluginName`. `_.omit()` does the job, but it isn't really
+    // type-safe.  so we're fudging it here.
+    delete appium[/** @type {string} */(`${this.type}Name`)];
+    return /** @type {ExtensionFields<ExtType>} */(/** @type {unknown} */({...appium, pkgName: name, version}));
```
this one's a mind bender

boneskull replied:
yeah. like, you can't just `delete` prop `foo` from `baz` of type `Bar` if `foo` is not optional.  var `baz` remains `Baz`, which is no longer true?  I don't know.  even a bare type casting didn't work, so you first cast to `unknown` then to whatever you want.

I suppose I could rewrite this by explicitly pulling out each prop we want from `appium` but that seems really tedious.

jlipps replied:
i'm kinda less concerned about type gymnastics here than the fact that I can't even see what is returned on one line because of the type comment.

boneskull replied:
I'll make it more legible

jlipps commented on packages/appium/lib/cli/extension.js:
```diff
@@ -30,7 +35,7 @@ async function runExtensionCommand (args, configObject) {
   let config = configObject;
   config.log = logFn;
-  const CommandClass = type === DRIVER_TYPE ? DriverCommand : PluginCommand;
+  const CommandClass = /** @type {ExtType extends import('../extension/manifest').DriverType ? import('@appium/types').Class<DriverCommand> : import('@appium/types').Class<PluginCommand>} */(commandClasses[type]);
```
can we put this in a typedef or something so it doesn't impact readability so much?

jlipps commented on packages/appium/lib/main.js:
```diff
@@ -1,7 +1,6 @@
 #!/usr/bin/env node
 
 // transpile:main
```
btw, with the removal of gulp, i think we can get rid of all of these `//transpile` comments?

jlipps commented on packages/appium/lib/utils.js:
```diff
@@ -1,4 +1,3 @@
-// @ts-check
```
what is the consequence of removing `// @ts-check` from files?

boneskull replied:
`// @ts-check` tells typescript "check this file" regardless if it is configured to check the file.  previously, typescript was _not_ configured to typecheck _any_ javascript, and we enabled it on a per-file basis instead.  now that is no longer true--at least for the `appium` package--as we've added `checkJs: true` to its `tsconfig.json`.  thus the `// @ts-check`s are redundant.

jlipps commented on packages/appium/types/cli.d.ts:
```diff
@@ -0,0 +1,137 @@
+ * Possible subcommands for the `appium` CLI.
+ */
+type CliSubcommand =
+  | typeof SERVER_SUBCOMMAND
```
weird, you can put a `|` as a no-op prefix operator?

jlipps commented on packages/appium/types/cli.d.ts:
```diff
@@ -0,0 +1,137 @@
+  /**
+   * Possible subcommands
+   */
+  subcommand:
```
how is this different from `CliSubcommand`?

boneskull replied:
It's not. good catch

boneskull commented:
this is failing b/c I changed something I thought was safe to change, but it wasn't.

boneskull commented on packages/appium/lib/appium.js:
```diff
@@ -21,48 +25,78 @@ const desiredCapabilityConstraints = {
+   * some commands are sessionless, so we need a set of plugins for them
+   * @type {InstanceType<import('../types').Extension.PluginClass>[]}
+   */
+  sessionlessPlugins = [];;
```
for some definition of 'unnecessary' ;)

boneskull committed:
chore(appium): remove transpile directives
this was ostensibly for gulp-babel

boneskull committed:
chore(appium): remove redundant type declaration

boneskull committed:
fix(appium): avoid potential unhandled rejections

- make types more readable

jlipps commented on packages/appium/lib/cli/extension-command.js:
```diff
@@ -237,6 +242,9 @@ export default class ExtensionCommand {
       extData = await this.installViaNpm({ext, pkgName, pkgVer});
     }
 
+    /**
```
jeez does this have to be 3 lines?

boneskull replied:
it doesn't need to exist at all

boneskull committed:
chore(appium): fix capabilities types

- move `scripts/generate*` elsewhere
- add reference to `@appium/types` in `tsconfig.json`

boneskull committed:
chore(appium): review updates for legibility
